src/registrar.js

```diff
@@ -0,0 +1,165 @@
+
+    for (const [, topology] of this.multicodecTopologies) {
+      topology.disconnect(peerInfo, error)
+    }
+
+    this.connections.delete(peerInfo.id.toB58String())
+  }
+
+  /**
+   * Get a connection with a peer.
+   * @param {PeerInfo} peerInfo
+   * @returns {Connection}
+   */
+  getPeerConnection (peerInfo) {
+    assert(PeerInfo.isPeerInfo(peerInfo), 'peerInfo must be an instance of peer-info')
+
+    return this.connections.get(peerInfo.id.toB58String())
+  }
+
+  /**
+   * Register handlers for a set of multicodecs given
+   * @param {Array<string>|string} multicodecs
+   * @param {object} handlers
+   * @param {function} handlers.onConnect
+   * @param {function} handlers.onDisconnect
+   * @param {object} topology properties for topology
+   * @return {string} registrar identifier
+   */
+  register (multicodecs, handlers, topology = {}) {
+    if (!multicodecs) {
+      throw errCode(new Error('one or more multicodec should be provided'), 'ERR_NO_MULTICODECS')
+    } else if (!Array.isArray(multicodecs)) {
+      multicodecs = [multicodecs]
+    }
+
+    if (!handlers) {
+      throw errCode(new Error('the handlers should be provided'), 'ERR_NO_HANDLERS')
+    } else if (!handlers.onConnect || typeof handlers.onConnect !== 'function') {
+      throw errCode(new Error('the \'onConnect\' handler must be provided'), 'ERR_NO_ONCONNECT_HANDLER')
+    } else if (!handlers.onDisconnect || typeof handlers.onDisconnect !== 'function') {
+      throw errCode(new Error('the \'onDisconnect\' handler must be provided'), 'ERR_NO_ONDISCONNECT_HANDLER')
+    }
+
+    // Create multicodec topology
+    const multicodecTopology = new MulticodecTopology({
+      onConnect: handlers.onConnect,
+      onDisconnect: handlers.onDisconnect,
+      registrar: this,
+      multicodecs,
+      ...topology
+    })
+
+    multicodecTopology.peerStore = this.peerStore
+
+    const id = (parseInt(Math.random() * 1e9)).toString(36) + Date.now()
+    this.multicodecTopologies.set(id, multicodecTopology)
+
+    this._addConnectedPeers(multicodecs, multicodecTopology)
+
+    // TODO: try to connect to peers-store peers according to current topology
+
+    return id
+  }
+
+  _addConnectedPeers (multicodecs, multicodecTopology) {
+    const knownPeers = this.peerStore.getAllArray()
+      .filter((peerInfo) => multicodecs.filter(multicodec => peerInfo.protocols.has(multicodec)))
+
+    for (const [id, conn] of this.connections.entries()) {
+      const targetPeer = knownPeers.find((peerInfo) => peerInfo.id.toB58String() === id)
+
+      if (targetPeer) {
+        multicodecTopology.tryToConnect(targetPeer, conn)
+      }
+    }
+  }
+
+  /**
+   * Unregister topology.
+   * @param {string} id registrar identifier
+   * @return {void}
+   */
+  unregister (id) {
+    const topology = this.multicodecTopologies.get(id)
+
+    if (!topology) {
+      throw errCode(new Error('no registrar found for the provided id'), 'ERR_NO_REGISTRAR')
```

Comment:
I don't think we need to throw here. If it's not being tracked, that's technically a successful unregistration imo.